File: flightnotify/services/quota.py
"""Free-tier quota enforcement.

The quota guard is a product behaviour, not documentation: normal application
activity cannot exceed the configured monthly allowance. Two independent
signals are combined and the *most conservative* wins:

1. a local ledger of every billable call FlightNotify made, and
2. the provider's own account status (free to query, per SerpApi's docs).

A configurable reserve is held back so a deliberate "Check now" still works
after automation has stopped.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import timedelta

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from ..config import Settings, get_settings
from ..enums import RunTrigger
from ..models import ProviderCall, ProviderUsage
from ..providers.base import FareProvider
from ..providers.errors import ProviderError
from ..timeutil import period_key, utcnow

log = logging.getLogger(__name__)

#: Triggers allowed to draw on the reserve. Automation never is.
RESERVE_ELIGIBLE_TRIGGERS = frozenset({RunTrigger.MANUAL, RunTrigger.INITIAL})


@dataclass(frozen=True, slots=True)
class QuotaSnapshot:
    period: str
    monthly_limit: int
    reserve: int
    local_used: int
    provider_used: int | None
    provider_left: int | None
    provider_limit: int | None
    provider_plan: str | None
    provider_account_masked: str | None
    last_synced_at: object | None
    sync_error: str | None
    hourly_limit: int
    hourly_used: int

    @property
    def effective_used(self) -> int:
        """The higher of our count and the provider's - never the lower."""
        if self.provider_used is None:
            return self.local_used
        return max(self.local_used, self.provider_used)

    @property
    def remaining_hard(self) -> int:
        """Calls left before the configured monthly cap, ignoring the reserve."""
        remaining = self.monthly_limit - self.effective_used
        if self.provider_left is not None:
            remaining = min(remaining, self.provider_left)
        return max(0, remaining)

    @property
    def remaining_safe(self) -> int:
        """Calls automation may still make."""
        return max(0, self.remaining_hard - self.reserve)

    @property
    def used_percent(self) -> float:
        if self.monthly_limit <= 0:
            return 100.0
        return min(100.0, round(self.effective_used / self.monthly_limit * 100, 1))

    @property
    def hourly_remaining(self) -> int:
        return max(0, self.hourly_limit - self.hourly_used)

    @property
    def is_exhausted(self) -> bool:
        return self.remaining_hard <= 0

    @property
    def automation_blocked(self) -> bool:
        return self.remaining_safe <= 0


@dataclass(frozen=True, slots=True)
class SpendDecision:
    allowed: bool
    granted: int
    reason: str

    @property
    def blocked(self) -> bool:
        return not self.allowed


class QuotaManager:
    def __init__(self, settings: Settings | None = None) -> None:
        self.settings = settings or get_settings()

    # -- ledger -------------------------------------------------------------
    def usage_row(self, session: Session, period: str | None = None) -> ProviderUsage:
        period = period or period_key()
        row = session.execute(
            select(ProviderUsage).where(
                ProviderUsage.provider == "serpapi", ProviderUsage.period == period
            )
        ).scalar_one_or_none()
        if row is None:
            row = ProviderUsage(provider="serpapi", period=period, local_searches=0)
            session.add(row)
            session.flush()
        return row

    def hourly_used(self, session: Session) -> int:
        since = utcnow() - timedelta(hours=1)
        total = session.execute(
            select(func.count(ProviderCall.id)).where(
                ProviderCall.provider == "serpapi", ProviderCall.called_at >= since
            )
        ).scalar_one()
        return int(total or 0)

    def snapshot(self, session: Session) -> QuotaSnapshot:
        row = self.usage_row(session)
        return QuotaSnapshot(
            period=row.period,
            monthly_limit=self.settings.serpapi_monthly_search_limit,
            reserve=self.settings.serpapi_reserve_searches,
            local_used=row.local_searches,
            provider_used=row.provider_this_month_usage,
            provider_left=row.provider_searches_left,
            provider_limit=row.provider_searches_per_month,
            provider_plan=row.provider_plan_name,
            provider_account_masked=row.provider_account_email_masked,
            last_synced_at=row.last_synced_at,
            sync_error=row.last_sync_error,
            hourly_limit=self.settings.serpapi_hourly_search_limit,
            hourly_used=self.hourly_used(session),
        )

    # -- decisions ----------------------------------------------------------
    def authorize(
        self, session: Session, *, wanted: int, trigger: RunTrigger | str
    ) -> SpendDecision:
        """Decide how many of ``wanted`` billable calls may be made now."""
        trigger = RunTrigger(trigger)
        snapshot = self.snapshot(session)
        may_use_reserve = trigger in RESERVE_ELIGIBLE_TRIGGERS
        budget = snapshot.remaining_hard if may_use_reserve else snapshot.remaining_safe

        if snapshot.monthly_limit <= 0:
            return SpendDecision(
                False, 0, "SERPAPI_MONTHLY_SEARCH_LIMIT is 0, so no searches are permitted."
            )
        if budget <= 0:
            if snapshot.remaining_hard <= 0:
                return SpendDecision(
                    False,
                    0,
                    "Monthly provider allowance is exhausted "
                    f"({snapshot.effective_used}/{snapshot.monthly_limit} used "
                    f"in {snapshot.period}).",
                )
            return SpendDecision(
                False,
                0,
                f"Only the {snapshot.reserve}-search reserve remains "
                f"({snapshot.remaining_hard} left); automated checks are paused so a "
                "manual check stays possible.",
            )

        hourly_left = snapshot.hourly_remaining
        if hourly_left <= 0:
            return SpendDecision(
                False,
                0,
                f"The provider's hourly throughput limit ({snapshot.hourly_limit}/hour) "
                "is reached. FlightNotify will resume on a later run.",
            )

        granted = min(wanted, budget, hourly_left)
        if granted < wanted:
            return SpendDecision(
                True,
                granted,
                f"Reduced from {wanted} to {granted} searches to stay within the "
                f"configured allowance ({snapshot.remaining_hard} left this period).",
            )
        return SpendDecision(True, granted, "")

    # -- recording ----------------------------------------------------------
    def record_call(
        self, session: Session, *, endpoint: str, run_id: int | None = None, count: int = 1
    ) -> None:
        """Record billable calls. Must be called for every non-cached search.

        FlightNotify counts a search that returned no itineraries too. SerpApi
        states errored searches are not billed, but over-counting stops
        automation early, which is the safe direction for a hard cap.
        """
        if count <= 0:
            return
        row = self.usage_row(session)
        row.local_searches += count
        for _ in range(count):
            session.add(
                ProviderCall(
                    provider="serpapi", endpoint=endpoint, called_at=utcnow(), search_run_id=run_id
                )
            )
        session.flush()

    def prune_call_log(self, session: Session, *, keep_hours: int = 72) -> int:
        """Trim the throughput log; the monthly ledger is authoritative."""
        cutoff = utcnow() - timedelta(hours=keep_hours)
        rows = (
            session.execute(select(ProviderCall).where(ProviderCall.called_at < cutoff))
            .scalars()
            .all()
        )
        for row in rows:
            session.delete(row)
        return len(rows)

    # -- provider sync ------------------------------------------------------
    def sync_from_provider(self, session: Session, provider: FareProvider) -> QuotaSnapshot:
        """Refresh provider-reported quota. Never consumes a fare search."""
        row = self.usage_row(session)
        if not provider.is_configured():
            row.last_sync_error = "No SerpApi key configured."
            session.flush()
            return self.snapshot(session)
        try:
            status = provider.account_status()
        except ProviderError as exc:
            row.last_sync_error = exc.guidance()
            session.flush()
            log.warning("provider quota sync failed", extra={"error_category": exc.category.value})
            return self.snapshot(session)

        row.provider_plan_name = status.plan_name
        row.provider_searches_per_month = status.searches_per_month
        row.provider_searches_left = status.searches_left
        row.provider_this_month_usage = status.this_month_usage
        row.provider_rate_limit_per_hour = status.rate_limit_per_hour
        row.provider_account_email_masked = status.account_email_masked
        row.last_synced_at = status.fetched_at
        row.last_sync_error = None
        # Never let the local ledger under-report what the provider has seen.
        if status.this_month_usage is not None:
            row.local_searches = max(row.local_searches, status.this_month_usage)
        session.flush()
        return self.snapshot(session)

File: worker/src/services/quota.ts
/**
 * Free-tier quota enforcement.
 *
 * Port of `flightnotify/services/quota.py`. The guard is product behaviour,
 * not documentation: normal application activity cannot exceed the configured
 * monthly allowance. Two independent signals are combined and the *most
 * conservative* wins -- a local ledger of every billable call, and the
 * provider's own reported account status.
 *
 * A configurable reserve is held back so a deliberate "Check now" still works
 * after automation has stopped.
 */

import type { Config } from "../env.js";
import type { Repo } from "../db/repo.js";
import { RunTrigger, type RunTriggerValue } from "../domain/enums.js";
import { periodKey } from "../time.js";

/** Triggers allowed to draw on the reserve. Automation never is. */
const RESERVE_ELIGIBLE: ReadonlySet<string> = new Set([RunTrigger.MANUAL, RunTrigger.INITIAL]);

export interface QuotaSnapshot {
  period: string;
  monthlyLimit: number;
  reserve: number;
  reservePercent: number;
  localUsed: number;
  providerUsed: number | null;
  providerLeft: number | null;
  providerLimit: number | null;
  providerPlan: string | null;
  providerAccountMasked: string | null;
  lastSyncedAt: string | null;
  syncError: string | null;
  hourlyLimit: number;
  hourlyUsed: number;

  effectiveUsed: number;
  remainingHard: number;
  remainingSafe: number;
  usedPercent: number;
  hourlyRemaining: number;
  isExhausted: boolean;
  automationBlocked: boolean;
}

export interface SpendDecision {
  allowed: boolean;
  granted: number;
  reason: string;
}

export interface QuotaReservation {
  period: string;
  hour: string;
  reserved: number;
}

function hourKey(now: Date): string {
  return now.toISOString().slice(0, 13);
}

export class QuotaManager {
  constructor(
    private readonly repo: Repo,
    private readonly config: Config,
  ) {}

  async snapshot(now = new Date()): Promise<QuotaSnapshot> {
    const period = periodKey(now);
    const row = await this.repo.usageRow(period);
    const hourlyUsed = await this.repo.hourlyUsed(new Date(now.getTime() - 3600_000));

    const monthlyLimit = this.config.monthlySearchBudget;
    const reserve = this.config.reserveSearches;
    const localUsed = row.local_searches;
    const providerUsed = row.provider_this_month_usage;
    const providerLeft = row.provider_searches_left;

    // Never the lower of the two: if the provider has seen more calls than we
    // recorded, the provider is right and we stop sooner.
    const effectiveUsed = providerUsed === null ? localUsed : Math.max(localUsed, providerUsed);

    let remainingHard = monthlyLimit - effectiveUsed;
    if (providerLeft !== null) remainingHard = Math.min(remainingHard, providerLeft);
    remainingHard = Math.max(0, remainingHard);

    const remainingSafe = Math.max(0, remainingHard - reserve);
    const hourlyLimit =
      row.provider_rate_limit_per_hour === null
        ? this.config.hourlySearchLimit
        : Math.min(this.config.hourlySearchLimit, row.provider_rate_limit_per_hour);
    const hourlyRemaining = Math.max(0, hourlyLimit - hourlyUsed);

    return {
      period,
      monthlyLimit,
      reserve,
      reservePercent: this.config.reservePercent,
      localUsed,
      providerUsed,
      providerLeft,
      providerLimit: row.provider_searches_per_month,
      providerPlan: row.provider_plan_name,
      providerAccountMasked: row.provider_account_email_masked,
      lastSyncedAt: row.last_synced_at,
      syncError: row.last_sync_error,
      hourlyLimit,
      hourlyUsed,
      effectiveUsed,
      remainingHard,
      remainingSafe,
      usedPercent:
        monthlyLimit <= 0
          ? 100
          : Math.min(100, Math.round((effectiveUsed / monthlyLimit) * 1000) / 10),
      hourlyRemaining,
      isExhausted: remainingHard <= 0,
      automationBlocked: remainingSafe <= 0,
    };
  }

  /**
   * Reserve capacity before a provider request.
   *
   * Monthly and hourly counters each use an atomic conditional write.  They
   * are deliberately conservative across the tiny gap between the two writes:
   * a killed invocation can leave capacity reserved, but cannot spend an
   * unreserved call.  Housekeeping/reconciliation can recover stale capacity;
   * protecting the owner's paid allowance takes precedence over availability.
   */
  async reserveCalls(
    count: number,
    trigger: RunTriggerValue,
    now = new Date(),
  ): Promise<{ decision: SpendDecision; reservation: QuotaReservation | null }> {
    if (count <= 0) {
      return {
        decision: { allowed: true, granted: 0, reason: "" },
        reservation: { period: periodKey(now), hour: hourKey(now), reserved: 0 },
      };
    }

    const period = periodKey(now);
    const row = await this.repo.usageRow(period);
    const mayUseReserve = RESERVE_ELIGIBLE.has(trigger);
    const configuredLimit = mayUseReserve
      ? this.config.monthlySearchBudget
      : Math.max(0, this.config.monthlySearchBudget - this.config.reserveSearches);
    const allowedLimit =
      row.provider_searches_per_month === null
        ? configuredLimit
        : Math.min(configuredLimit, row.provider_searches_per_month);

    if (!(await this.repo.reserveMonthlyCalls(period, count, allowedLimit))) {
      const snapshot = await this.snapshot(now);
      return {
        reservation: null,
        decision: {
          allowed: false,
          granted: 0,
          reason:
            snapshot.remainingHard <= 0
              ? `Monthly provider allowance is exhausted (${snapshot.effectiveUsed}/${snapshot.monthlyLimit} used in ${snapshot.period}).`
              : mayUseReserve
                ? "The remaining provider allowance is smaller than this request's retry reservation."
                : `Only the ${snapshot.reserve}-search reserve remains; automated checks are paused.`,
        },
      };
    }

    const providerHourly = row.provider_rate_limit_per_hour;
    const hourlyLimit =
      providerHourly === null
        ? this.config.hourlySearchLimit
        : Math.min(this.config.hourlySearchLimit, providerHourly);
    const hour = hourKey(now);
    if (!(await this.repo.reserveHourlyCalls(hour, count, hourlyLimit))) {
      await this.repo.releaseMonthlyCalls(period, count);
      return {
        reservation: null,
        decision: {
          allowed: false,
          granted: 0,
          reason:
            `The provider's hourly throughput limit (${hourlyLimit}/hour) is reached. ` +
            "FlightNotify will resume on a later run.",
        },
      };
    }

    return {
      decision: { allowed: true, granted: count, reason: "" },
      reservation: { period, hour, reserved: count },
    };
  }

  /** Reconcile a conservative reservation with the adapter's actual attempts. */
  async finalizeReservation(
    reservation: QuotaReservation,
    endpoint: string,
    runId: number | null,
    actualCount: number,
    now = new Date(),
  ): Promise<void> {
    const actual = Math.max(0, Math.min(reservation.reserved, Math.trunc(actualCount)));
    const unused = reservation.reserved - actual;
    if (unused > 0) {
      await this.repo.releaseMonthlyCalls(reservation.period, unused);
      await this.repo.releaseHourlyCalls(reservation.hour, unused);
    }
    await this.repo.insertProviderCallRows(endpoint, runId, actual, now.toISOString());
  }

  /** Decide how many of `wanted` billable calls may be made now. */
  async authorize(
    wanted: number,
    trigger: RunTriggerValue,
    now = new Date(),
  ): Promise<{ decision: SpendDecision; snapshot: QuotaSnapshot }> {
    const snapshot = await this.snapshot(now);
    const mayUseReserve = RESERVE_ELIGIBLE.has(trigger);
    const budget = mayUseReserve ? snapshot.remainingHard : snapshot.remainingSafe;

    if (snapshot.monthlyLimit <= 0) {
      return {
        snapshot,
        decision: {
          allowed: false,
          granted: 0,
          reason: "MONTHLY_SEARCH_BUDGET is 0, so no searches are permitted.",
        },
      };
    }

    if (budget <= 0) {
      const reason =
        snapshot.remainingHard <= 0
          ? `Monthly provider allowance is exhausted (${snapshot.effectiveUsed}/${snapshot.monthlyLimit} used in ${snapshot.period}).`
          : `Only the ${snapshot.reserve}-search reserve remains (${snapshot.remainingHard} left); ` +
            "automated checks are paused so a manual check stays possible.";
      return { snapshot, decision: { allowed: false, granted: 0, reason } };
    }

    if (snapshot.hourlyRemaining <= 0) {
      return {
        snapshot,
        decision: {
          allowed: false,
          granted: 0,
          reason:
            `The provider's hourly throughput limit (${snapshot.hourlyLimit}/hour) is reached. ` +
            "FlightNotify will resume on a later run.",
        },
      };
    }

    const granted = Math.min(wanted, budget, snapshot.hourlyRemaining);
    if (granted < wanted) {
      return {
        snapshot,
        decision: {
          allowed: true,
          granted,
          reason:
            `Reduced from ${wanted} to ${granted} searches to stay within the configured ` +
            `allowance (${snapshot.remainingHard} left this period).`,
        },
      };
    }
    return { snapshot, decision: { allowed: true, granted, reason: "" } };
  }

  /**
   * Record billable calls.
   *
   * A search that returned no itineraries still counts. SerpApi states errored
   * searches are not billed, but over-counting stops automation early, which is
   * the safe direction for a hard cap on someone's bill.
   */
  async recordCalls(
    endpoint: string,
    runId: number | null,
    count: number,
    now = new Date(),
  ): Promise<void> {
    if (count <= 0) return;
    const period = periodKey(now);
    await this.repo.usageRow(period);
    await this.repo.recordProviderCalls(period, endpoint, runId, count);
  }

  /** Refresh provider-reported quota. Never consumes a fare search. */
  async syncFromProvider(
    status: {
      planName: string | null;
      searchesPerMonth: number | null;
      searchesLeft: number | null;
      thisMonthUsage: number | null;
      rateLimitPerHour: number | null;
      accountEmailMasked: string | null;
      fetchedAt: string;
    } | null,
    error: string | null,
    now = new Date(),
  ): Promise<void> {
    const period = periodKey(now);
    const row = await this.repo.usageRow(period);

    if (status === null) {
      await this.repo.updateUsage(period, { last_sync_error: error });
      return;
    }

    await this.repo.updateUsage(period, {
      provider_plan_name: status.planName,
      provider_searches_per_month: status.searchesPerMonth,
      provider_searches_left: status.searchesLeft,
      provider_this_month_usage: status.thisMonthUsage,
      provider_rate_limit_per_hour: status.rateLimitPerHour,
      provider_account_email_masked: status.accountEmailMasked,
      last_synced_at: status.fetchedAt,
      last_sync_error: null,
      // Never let the local ledger under-report what the provider has seen.
      local_searches:
        status.thisMonthUsage === null
          ? row.local_searches
          : Math.max(row.local_searches, status.thisMonthUsage),
    });
  }
}
